### PR TITLE
New feature

### PR DESCRIPTION
**Pull request by samuelvanderwaal**
Adds the `freeze` feature with the following commands:

* freeze enable -- enables the freeze feature for a candy machine that has not started minting yet
* freeze disable -- removes the freeze feature 
* thaw -- thaws a single frozen NFT if freeze rules allow it, or all NFTs from a candy machine with the `--all` option
* unfreeze-funds -- unlocks treasury funds if freeze rules allow it

### CODE REVIEW

File: src/freeze/disable.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        style("[2/2]").bold().dim(),
+        FIRE_EMOJI,
+        RIGHT_ARROW_EMOJI,
+        ICE_CUBE_EMOJI
+    );
+
+    let pb = spinner_with_style();
+    pb.set_message("Sending remove freeze transaction...");
+
+    let signature = disable_freeze(&program, &candy_pubkey)?;
```
Comment (febo) [inline src/freeze/disable.rs:57]: Is it possible to check whether the candy machine has freeze enabled before trying to disable? Or perhaps improve the error message?
image

Comment (febo) [reply src/freeze/disable.rs:57]: I am also not sure about the 3 emojis – I understand the concept, but they look a bit out of place compared to the other ones.

File: src/freeze/enable.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            cache.program.candy_machine
+        }
+    };
+
+    let candy_pubkey = Pubkey::from_str(&candy_machine_id)
+        .map_err(|_| anyhow!("Failed to parse candy machine id: {}", &candy_machine_id))?;
+
+    // Freeze days specified takes precedence over the one from the config.
+    let freeze_time = if let Some(freeze_days) = args.freeze_days {
+        (freeze_days * 24 * 60 * 60) as i64
```
Comment (febo) [inline src/freeze/enable.rs:32]: This line generated a panic for me:
image
My guess is that it was trying to do the multiplication as `u8`. I changed to `(freeze_days as i64) * 24 * 60 * 60` and it worked.

File: src/freeze/enable.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    println!(
+        "\n{} {}Turning on freeze feature for candy machine",
+        style("[2/2]").bold().dim(),
+        ICE_CUBE_EMOJI
+    );
+
+    let pb = spinner_with_style();
+    pb.set_message("Sending Enable Freeze transaction...");
+
+    let signature = enable_freeze(&program, &config_data, &candy_pubkey, freeze_time)?;
```
Comment (febo) [inline src/freeze/enable.rs:74]: Similar to the disable, trying to enable an already enabled freeze candy machine gives an unclear error:
image

File: src/freeze/enable.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            let cache = load_cache(&args.cache, false)?;
+            cache.program.candy_machine
+        }
+    };
+
+    let candy_pubkey = Pubkey::from_str(&candy_machine_id)
+        .map_err(|_| anyhow!("Failed to parse candy machine id: {}", &candy_machine_id))?;
+
+    // Freeze days specified takes precedence over the one from the config.
+    let freeze_time = if let Some(freeze_days) = args.freeze_days {
```
Comment (febo) [inline src/freeze/enable.rs:31]: Perhaps we should validate that `args.freeze_days` is in a valid range (`0 < freeze_days <= 30`):
image

File: src/freeze/enable.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    pb.finish_with_message("Done");
+
+    assert_correct_authority(
+        &sugar_config.keypair.pubkey(),
+        &candy_machine_state.authority,
+    )?;
+
+    // Cannot enable freeze if minting has started.
+    if candy_machine_state.items_redeemed > 0 {
+        return Err(anyhow!("Cannot enable freeze after minting has started"));
```
Comment (febo) [inline src/freeze/enable.rs:62]: This is a question: you can disable freeze after minting started but you cannot enable it?

File: src/freeze/thaw.rs [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+            .unwrap()
+            .value
+            .unwrap();
+        let account_data = SplAccount::unpack(&account.data).unwrap();
+        let owner = account_data.owner;
+
+        // Only thaw frozen accounts.
+        if !account_data.is_frozen() {
+            println!("\n NFT is already thawed! NFT");
+            std::process::exit(0);
```
Comment (febo) [inline src/freeze/thaw.rs:145]: Change this line for `return Ok(());` so the command exits gracefully.

File: src/freeze/thaw.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+            .get_account_with_commitment(&token_account, CommitmentConfig::confirmed())
+            .unwrap()
+            .value
+            .unwrap();
+        let account_data = SplAccount::unpack(&account.data).unwrap();
+        let owner = account_data.owner;
+
+        // Only thaw frozen accounts.
+        if !account_data.is_frozen() {
+            println!("\n NFT is already thawed! NFT");
```
Comment (febo) [inline src/freeze/thaw.rs:144]: Perhaps remove the extra ` NFT` part of the message at the end.

File: src/freeze/unfreeze_funds.rs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    println!(
+        "\n{} {}Unlocking treasury funds. . .",
+        style("[2/2]").bold().dim(),
+        MONEY_BAG_EMOJI
+    );
+
+    let pb = spinner_with_style();
+    pb.set_message("Sending unlock funds transaction...");
+
+    let signature = unlock_funds(&program, &candy_pubkey)?;
```
Comment (febo) [inline src/freeze/unfreeze_funds.rs:55]: When I try to unlock funds of a candy machine that does not have lock funds a get the following:
image

File: src/freeze/enable.rs (hunk 5 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:62]: Correct.

File: src/freeze/enable.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            pubkey: freeze_ata,
+            is_signer: false,
+            is_writable: true,
+        };
+        additional_accounts.push(freeze_ata_account);
+
+        builder = builder.instruction(freeze_ata_ix);
+    }
+
+    println!("Program payer: {}", program.payer());
```
Comment (febo) [inline src/freeze/enable.rs:112]: Is this a debug message?

File: src/freeze/unfreeze_funds.rs (hunk 8 above)
Comment (febo) [inline src/freeze/unfreeze_funds.rs:55]: Trying to unfreeze a candy machine that is not "froze" gets this error:
image

File: src/freeze/enable.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
+use super::*;
+
+pub struct EnableFreezeArgs {
+    pub keypair: Option<String>,
+    pub rpc_url: Option<String>,
+    pub cache: String,
+    pub config: String,
+    pub candy_machine: Option<String>,
+    pub freeze_days: Option<u8>,
```
Comment (febo) [inline src/freeze/enable.rs:9]: Perhaps we need to have the `freeze_days` under a flag or as the first argument. I tried to specify the `--cache` and the `freeze_days` but the command tried to parse the argument as the candy machine id:
image

File: src/freeze/disable.rs (hunk 1 above)
Comment (samuelvanderwaal) [reply src/freeze/disable.rs:64]: [This commit](https://github.com/metaplex-foundation/sugar/pull/345/commits/6bdb79ff9f6e5b53a1014497116eb5ec45a2b478) fixes these two issues.

Comment (samuelvanderwaal) [reply src/freeze/disable.rs:64]: It also changes the candy machine to be an option for consistency and so a user doesn't have to provide it if they provide a freeze time.

File: src/freeze/enable.rs (hunk 2 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs]: [Fixed](https://github.com/metaplex-foundation/sugar/pull/345/commits/566258c4317e8301d636b1c47eead7d2dc8ac4e6).

Selection_961

File: src/freeze/enable.rs (hunk 3 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:83]: [Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/a4d09052c789a0156cc8cd9f3fd3ed4d7d0235ff)

Selection_962

File: src/freeze/enable.rs (hunk 4 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:31]: Should this take `freezeTime` in seconds instead? It seems like taking days is a better UI and advanced users can still set seconds in the config file. 

I'll add a check for `freeze_days > 30` which is the only range we need to validate given it's an unsigned integer.

File: src/freeze/enable.rs (hunk 5 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:71]: Candy machine limitation, not Sugar.

File: src/freeze/enable.rs (hunk 4 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:31]: Selection_963

File: src/freeze/enable.rs (hunk 9 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs]: Yes, good catch.

[Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/274e3759ac42808a82f4e05dd5a0fa7ab591961f)

File: src/freeze/thaw.rs (hunk 7 above)
Comment (samuelvanderwaal) [reply src/freeze/thaw.rs]: [Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/274e3759ac42808a82f4e05dd5a0fa7ab591961f)

File: src/freeze/thaw.rs (hunk 6 above)
Comment (samuelvanderwaal) [reply src/freeze/thaw.rs]: [Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/274e3759ac42808a82f4e05dd5a0fa7ab591961f)

File: src/freeze/unfreeze_funds.rs (hunk 8 above)
Comment (samuelvanderwaal) [reply src/freeze/unfreeze_funds.rs:64]: [Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/274e3759ac42808a82f4e05dd5a0fa7ab591961f)

cilantro@pop-os: ~-Sync-sugar_test2_001

File: src/freeze/enable.rs (hunk 10 above)
Comment (samuelvanderwaal) [reply src/freeze/enable.rs:9]: [Fixed.](https://github.com/metaplex-foundation/sugar/pull/345/commits/6bdb79ff9f6e5b53a1014497116eb5ec45a2b478)


cilantro@pop-os: ~-Sync-sugar_test2_002

File: src/freeze/disable.rs (hunk 1 above)
Comment (febo) [reply src/freeze/disable.rs:64]: Great!

**Review by febo — APPROVED** [review]

**Pull request merged**